File: src/documents/draft-store.ts
// Recovery-draft persistence (audit §27–28). P0 implementation: debounced
// localStorage, keyed by document id — works identically inside the Tauri
// WebView and a browser tab with no runtime branching. This is a known,
// disclosed simplification of the audit's fuller recommendation (Tauri
// app-data dir / OPFS): localStorage has a ~5-10MB per-origin quota and is
// cleared by "clear site data", which is an acceptable P0 risk for typical
// Markdown file sizes but not a P1-complete implementation — see
// progress.md's P1 item for the follow-up.

export interface RecoveryDraft {
  documentId: string
  displayName: string
  text: string
  baseModifiedAt?: number
  draftModifiedAt: number
}

const KEY_PREFIX = 'claymark:draft:'

function keyFor(documentId: string): string {
  // documentId can contain arbitrary characters (a full path or content://
  // URI) — hash-free but namespaced; localStorage keys have no character
  // restriction, so no encoding is needed beyond the prefix.
  return `${KEY_PREFIX}${documentId}`
}

export const draftStore = {
  put(draft: RecoveryDraft): void {
    try {
      localStorage.setItem(keyFor(draft.documentId), JSON.stringify(draft))
    } catch {
      // Quota exceeded or storage disabled — the draft is best-effort; the
      // explicit Save path remains the real persistence guarantee.
    }
  },

  get(documentId: string): RecoveryDraft | null {
    try {
      const raw = localStorage.getItem(keyFor(documentId))
      return raw ? (JSON.parse(raw) as RecoveryDraft) : null
    } catch {
      return null
    }
  },

  remove(documentId: string): void {
    try {
      localStorage.removeItem(keyFor(documentId))
    } catch {
      // no-op
    }
  },
}

File: claymark-native/app/src/main/java/com/claymark/nativeapp/documents/DraftStore.kt
package com.claymark.nativeapp.documents

import android.content.Context
import org.json.JSONObject

/**
 * Port of `src/documents/draft-store.ts`.
 *
 * The web build's disclosed P0 simplification was localStorage: a ~5–10 MB
 * per-origin quota, cleared by "clear site data". SharedPreferences has the
 * same shape (a synchronous string map, cleared by "Clear storage") and the
 * same caveat, so this is a like-for-like port rather than an upgrade — the
 * explicit Save path remains the real persistence guarantee.
 */
data class RecoveryDraft(
    val documentId: String,
    val displayName: String,
    val text: String,
    val baseModifiedAt: Long? = null,
    val draftModifiedAt: Long,
)

private const val KEY_PREFIX = "claymark:draft:"

class DraftStore(context: Context) {
    private val prefs = context.getSharedPreferences("claymark-drafts", Context.MODE_PRIVATE)

    private fun keyFor(documentId: String) = "$KEY_PREFIX$documentId"

    fun put(draft: RecoveryDraft) {
        try {
            val json = JSONObject()
                .put("documentId", draft.documentId)
                .put("displayName", draft.displayName)
                .put("text", draft.text)
                .put("draftModifiedAt", draft.draftModifiedAt)
            draft.baseModifiedAt?.let { json.put("baseModifiedAt", it) }
            prefs.edit().putString(keyFor(draft.documentId), json.toString()).apply()
        } catch (_: Exception) {
            // Quota exceeded or storage disabled — the draft is best-effort.
        }
    }

    fun get(documentId: String): RecoveryDraft? = try {
        prefs.getString(keyFor(documentId), null)?.let { raw ->
            val json = JSONObject(raw)
            RecoveryDraft(
                documentId = json.getString("documentId"),
                displayName = json.optString("displayName"),
                text = json.getString("text"),
                baseModifiedAt = if (json.has("baseModifiedAt")) json.getLong("baseModifiedAt") else null,
                draftModifiedAt = json.optLong("draftModifiedAt"),
            )
        }
    } catch (_: Exception) {
        null
    }

    fun remove(documentId: String) {
        try {
            prefs.edit().remove(keyFor(documentId)).apply()
        } catch (_: Exception) {
            // no-op
        }
    }
}
